File: quattro2/win10/RolandWirelessConnect/RWCNetClient.cpp
/*
 * @(#)RWCNetClient.cpp
 *
 * Copyright 2013 Roland Corporation, Japan. All rights reserved.
 */

#include "RWCNetClient.h"
#include "RWCProtocol.h"

/* class KeepAlive */

class KeepAlive : public Thread
{
  public:
	KeepAlive(RWCNetClient* client) :
		mp_client(client), m_repeat(true), m_counter(0) { }
	virtual ~KeepAlive() { }
	void touch() { m_counter = 0; }
	void stop() { m_repeat = false; }

  private:
	RWCNetClient* mp_client;
	bool m_repeat;
	int m_counter;
	void run();
};

void KeepAlive::run()
{
	DWORD t0 = Timer::currentMSecTimestamp();
	while (m_repeat) {
		::Sleep(100);
		DWORD t = Timer::currentMSecTimestamp();
		if ((t - t0) < 1000) continue;
		t0 = t;

		int timeout = mp_client->keepAliveTimeout();

		if (mp_client->connecting() && timeout > 0) {
			RNP_HEADER header;
			header.type    = RNP_TYPE_KEEP_ALIVE;
			header.subtype = RNP_SUBTYPE_0;
			header.size[0] = 0;
			header.size[1] = 0;
			mp_client->send((BYTE*)&header, sizeof(RNP_HEADER));
			if (++m_counter > timeout) {
				mp_client->kill();
				break;
			}
		}
	}
}

/* class RWCNetClient */

RWCNetClient::RWCNetClient()
{
	mp_socket = 0;

	m_acceptable = false;
	m_connecting = false;

	m_connectTimeout = kDefaultConnectTimeout;
	m_keepAliveTimeout = kDefaultKeepAliveTimeout;
}

void RWCNetClient::connectToServer(LPCTSTR address, const int portNo)
{
	mutex_lock obj(m_mtx);

	if ((mp_socket = new Socket()) && mp_socket->create()) {
		m_acceptable = true;
		_address = inet_addr(CStringA(address));
		_portNo  = portNo;
		_threadStart(); /* call run() on new Thread */
	} else {
		delete mp_socket;
		mp_socket = 0;
		streamOpenFailed();
	}
}

void RWCNetClient::stopConnection()
{
	mutex_lock obj(m_mtx);

	m_acceptable = false;

	if (mp_socket) {
		mp_socket->close();
		wait();
		delete mp_socket;
		mp_socket = 0;
	}
}

int RWCNetClient::send(BYTE* buffer, int length)
{
	mutex_lock obj(m_mtx);

	return mp_socket ? mp_socket->send(buffer, length) : 0;
}

void RWCNetClient::run()
{
	if (!mp_socket->connect(_address, _portNo, m_connectTimeout * 1000)) {
		mp_socket->close();
		if (m_acceptable) {
			m_acceptable = false;
			streamOpenFailed();
		}
		return;
	}

	m_connecting = true;
	streamOpenCompleted();

	KeepAlive keepAlive(this);
	keepAlive._threadStart();

	while (m_acceptable && streamRunLoop(mp_socket)) {
		keepAlive.touch();
	}
	keepAlive.stop();

	mp_socket->close();

	m_connecting = false;
	if (m_acceptable) {
		m_acceptable = false;
		streamEndEncountered();
	}

	keepAlive.wait();
}

void RWCNetClient::keepAliveTimeout(int seconds)
{
	m_keepAliveTimeout = seconds;

	BYTE buf[16];

	RNP_HEADER *header = (RNP_HEADER *)buf;
	header->type	= RNP_TYPE_KEEP_ALIVE;
	header->subtype = RNP_SUBTYPE_SET_TIMEOUT;
	header->size[0] = 0;
	header->size[1] = 1;

	RNP_KEEP_ALIVE_TIMEOUT *timeout = (RNP_KEEP_ALIVE_TIMEOUT *)(buf + sizeof(RNP_HEADER));
	timeout->seconds = (unsigned char)seconds;

	send(buf, sizeof(RNP_HEADER) + 1);
}

void RWCNetClient::kill()
{
	mutex_lock obj(m_mtx);

	if (mp_socket) {
		mp_socket->close();
	}
}

File: quattro2/android/main/java/quattro/WirelessConnect/RWCNetClient.java
//
//	RWCInputStream.java
//
//	Copyright 2013 Roland Corporation. All rights reserved.
//

package quattro.WirelessConnect;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;

abstract class RWCNetClient {

	abstract void streamOpenCompleted();
	abstract void streamOpenFailed();
	abstract void streamErrorOccurred();
	abstract void streamEndEncountered();

	abstract boolean streamRunLoop(DataInputStream in);

	protected int connectTimeout = RWCProtocol.kDefaultConnectTimeout;
	protected int keepAliveTimeout = RWCProtocol.kDefaultKeepAliveTimeout;

	public int connectTimeout() {
		return connectTimeout;
	}

	public void connectTimeout(int sec) {
		connectTimeout = sec;
	}

	public int keepAliveTimeout() {
		return keepAliveTimeout;
	}

	public void keepAliveTimeout(int sec) {
		keepAliveTimeout = sec;
		send(RWCProtocol.toKeepAliveTimeout(sec));
	}

	private InetSocketAddress address = null;
	private Socket socket = null;
	private Connection connection = null;
	private boolean acceptable = false;
	private boolean connecting = false;

	synchronized protected void connectToServer(String addr, int port) {
		acceptable = true;
		address = new InetSocketAddress(addr, port);
		socket = new Socket();
		connection = new Connection();
		connection.start();
	}

	synchronized protected void stopConnection() {
		acceptable = false;
		if (connection != null) {
			try {
				socket.close();
			} catch (IOException e) {}
			try {
				connection.join();
			} catch (InterruptedException e) {}
			socket = null;
			connection = null;
		}
	}

	synchronized void send(byte[] data) {
		if (connection != null) {
			connection.send(data);
		}
	}

	private class Connection extends Thread {

		private DataInputStream in = null;
		private DataOutputStream out = null;

		@Override
		public void run() {

			try {
				socket.connect(address, connectTimeout * 1000);
				in = new DataInputStream(socket.getInputStream());
				out = new DataOutputStream(socket.getOutputStream());
			} catch (IOException e) {
				streamOpenFailed();
				return;
			}

			connecting = true;
			streamOpenCompleted();

			KeepAlive keepAlive = new KeepAlive(this);
			keepAlive.start();

			while (acceptable && streamRunLoop(in)) {
				keepAlive.touch();
			}

			keepAlive.kill();

			connecting = false;
			if (acceptable) {
				acceptable = false;
				streamEndEncountered();
			}

			try {
				keepAlive.join();
			} catch (InterruptedException e) {}

			in = null;
			out = null;
			try {
				socket.close();
			} catch (IOException e) {}
		}

		synchronized void send(byte[] data) {
			if (out != null && data != null) {
				try {
					out.write(data);
				} catch (IOException e) {}
			}
		}
	}

	private class KeepAlive extends Thread {

		private Connection conn = null;
		private boolean repeat = true;
		private int counter = 0;

		public KeepAlive(Connection c) {
			conn = c;
		}
		public void touch() {
			counter = 0;
		}
		public void kill() {
			repeat = false;
		}

		@Override
		public void run() {
			long t0 = System.currentTimeMillis();
			while (repeat) {
				try {
					Thread.sleep(100);
				} catch (InterruptedException e) {}
				long t = System.currentTimeMillis();
				if ((t - t0) < 1000) continue;
				t0 = t;

				int timeout = keepAliveTimeout;
				if (connecting && timeout > 0) {
					conn.send(RWCProtocol.toKeepAlive());
					if (++counter > timeout) {
						try {
							socket.close();
						} catch (IOException e) {}
						break;
					}
				}
			}
			conn = null;
		}
	}

}
